### DISCUSSION TITLE
no upgrade path from the database dbscheme (814fae59) to the target dbscheme (ddd31fd0)

### DISCUSSION THREAD

**Discussion by timimin** [github/codeql#14939] [General]
Hello all

I'm trying to execute https://securitylab.github.com/ctf/uboot/

The problem concerns the pre-existing vulnerable uboot CodeQL database

https://github.com/github/securitylab/releases/download/u-boot-codeql-database/u-boot_u-boot_cpp-srcVersion_d0d07ba86afc8074d79e436b1ba4478fa0f0c1b5-dist_odasa-2019-07-25-linux64.zip

When i try to execute any query it produces the following error

```
CLI command succeeded.

Loaded database u-boot_u-boot_d0d07ba at URI file:///c:/codeql/codeql/u-boot_db/u-boot_u-boot_d0d07ba.
Finished loading persisted databases.
Registering jump-to-definition handlers.
Removing orphaned databases from workspace storage.
No orphaned databases found.
Reading query history
Reading cached query history from 'c:\Users\Admin\AppData\Roaming\Code\User\workspaceStorage\d480d41fc085e8bdbd4e105b5fe3c867\GitHub.vscode-codeql\workspace-query-history.json'.
Successfully finished extension initialization.
Resolving query metadata using CodeQL CLI: resolve metadata -v --log-to-stderr --format json c:\codeql\codeql\codeql-uboot\3_function_definitions.ql...
CLI command succeeded.

Failed to run query: There was no upgrade path from the database dbscheme (814fae59) to the target dbscheme (ddd31fd0)
```

Seems that the dbscheme of the pre-existing database doesn't fit.
creating a new database is problematic since  codeql database create --language=cpp ... doesn't work by default for this project...

upgrading QL dataset with

codeql resolve upgrades --dbscheme=<file> <options>...

has no effect.

Thanks in advance. Alex.

**Comment by jketema** [comment]
It seems that the database pre-dates the introduction of upgrade scripts to newer database scheme versions. Given that there is no available download for CodeQL (or Odasa, the precursor to CodeQL) that dates back to July 2019, your only option will be to recreate the database from scratch.

> creating a new database is problematic since codeql database create --language=cpp ... doesn't work by default for this project...

Note that you can specify the build command that should be used with the `--command` option.

**Reply by timimin** [reply]
Thank you very much

**Comment by timimin** [comment]
Sorry, but recreating the database didn't solve the problem.

I created a new database:

```
sudo apt-get install flex
sudo apt install libsdl1.2-dev
make 3c120_defconfig

~/Downloads/codeql-linux64/codeql/codeql database create cpp-databse --language=cpp --overwrite
```
 
then replaced the old db with the new one in my project
but the problem remains ...
```

Failed to run query: There was no upgrade path from the database dbscheme (8cba93a4) to the target dbscheme (ddd31fd0)
CodeQL extension version: 1.9.1 
CodeQL CLI version: 2.15.3 
Platform: win32 x64
```
I checked that codeql cli version:2.15.3 was used for database creation

**Reply by jketema** [reply]
It looks like you didn't properly replace the database. I'm not sure how you're replacing it, but I'd recommend getting rid of any old copy of the database before doing anything else, and for good measure not using the `--overwrite` option when creating the database.

**Reply by timimin** [reply]
Oh oh, i'm using Visual Studio Code extension. And there is an option: Choose database from Folder. --overwrite option was used for a new database directory (on another machine) to make it possible attempting `codeql database create ` command several times.

**Reply by jketema** [reply]
Ok, `8cba93a4` is actually a slightly more reasonable hash. It does look like you're using a different version of CodeQL to create the database than to run the query with.

Also what does your VSCode workspace look like, it is a copy of the starter workspace, if so, is the CodeQL submodule up-to-date?

**Reply by timimin** [reply]
Right. Updating CodeQL submodule helped. Thanks now it works ).

**Reply by jketema** [reply]
Glad that help. I'd still recommend building the database with the same version as you analyse it with, to profit from everything we might have fixed/improved.
